File: 2D_array/Matrix.java


public class Matrix {
    int[][] matrix = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    public void print() {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }

    public void transpose() {
        int[][] transposed = new int[matrix[0].length][matrix.length];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                transposed[j][i] = matrix[i][j];
            }
        }
        matrix = transposed;
    }

    public void rotateImage90Degrees() {
        transpose();
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length / 2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][matrix[i].length - 1 - j];
                matrix[i][matrix[i].length - 1 - j] = temp;
            }
        }
    }

    public void rotateImage180Degrees() {
        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int ni = rows - 1 - i;
                int nj = cols - 1 - j;
                if (i > ni || (i == ni && j >= nj)) {
                    continue;
                }
                int temp = matrix[i][j];
                matrix[i][j] = matrix[ni][nj];
                matrix[ni][nj] = temp;
            }
        }
    }

    public void spiralPrint() {
        int top = 0;
        int bottom = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;

        while (top <= bottom && left <= right) {
            for (int i = left; i <= right; i++) {
                System.out.print(matrix[top][i] + " ");
            }
            top++;

            for (int i = top; i <= bottom; i++) {
                System.out.print(matrix[i][right] + " ");
            }
            right--;

            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    System.out.print(matrix[bottom][i] + " ");
                }
                bottom--;
            }

            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    System.out.print(matrix[i][left] + " ");
                }
                left++;
            }
        }
    }

    public int[][] prefixSum() {
        int[][] prefSum = new int[matrix.length][matrix[0].length];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                prefSum[i][j] = matrix[i][j];
                if (i > 0) {
                    prefSum[i][j] += prefSum[i - 1][j];
                }
                if (j > 0) {
                    prefSum[i][j] += prefSum[i][j - 1];
                }
                if (i > 0 && j > 0) {
                    prefSum[i][j] -= prefSum[i - 1][j - 1];
                }
            }
        }
        return prefSum;
    }

    public static void main(String[] args) {
        Matrix m = new Matrix();
        m.print();

        System.out.print("\nSpiral Print: ");
        m.spiralPrint();

        System.out.println("\n\nTransposed Matrix:");
        m.transpose();
        m.print();

        System.out.println("\nPrefix Sum Matrix:");
        int[][] prefixSum = m.prefixSum();
        for (int i = 0; i < prefixSum.length; i++) {
            for (int j = 0; j < prefixSum[i].length; j++) {
                System.out.print(prefixSum[i][j] + " ");
            }
            System.out.println();
        }

        m.rotateImage90Degrees();
        System.out.println("\n90-Degree Rotated Matrix:");
        m.print();
        
        Matrix m2 = new Matrix();
        m2.rotateImage180Degrees();
        System.out.println("\n180-Degree Rotated Matrix:");
        m2.print();


    }
}

File: 2D_array/matrix.c
#include <stdio.h>

int matrix[3][3] = {
    {1, 2, 3},
    {4, 5, 6},
    {7, 8, 9}   
};

void print_matrix(int rows, int cols, int matrix[rows][cols]) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            printf("%d ", matrix[i][j]);
        }
        printf("\n");
    }
}

void transpose(int rows, int cols, int matrix[rows][cols]) {
    int temp;
    for (int i = 0; i < rows; i++) {
        for (int j = i + 1; j < cols; j++) {
            temp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = temp;
        }
    }
}

void rotate_90(int rows, int cols, int matrix[rows][cols]) {
    transpose(rows, cols, matrix);
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols / 2; j++) {
            int temp = matrix[i][j];
            matrix[i][j] = matrix[i][cols - j - 1];
            matrix[i][cols - j - 1] = temp;
        }
    }
}

void rotate_180(int rows, int cols, int matrix[rows][cols]) {
    rotate_90(rows, cols, matrix);
    rotate_90(rows, cols, matrix);
}

void spiral_print(int rows, int cols, int matrix[rows][cols]) {
    int top = 0, bottom = rows - 1, left = 0, right = cols - 1;
    while (top <= bottom && left <= right) {
        for (int i = left; i <= right; i++) {
            printf("%d ", matrix[top][i]);
        }
        top++;
        for (int i = top; i <= bottom; i++) {
            printf("%d ", matrix[i][right]);
        }
        right--;
        if (top <= bottom) {
            for (int i = right; i >= left; i--) {
                printf("%d ", matrix[bottom][i]);
            }
            bottom--;
        }
        if (left <= right) {
            for (int i = bottom; i >= top; i--) {
                printf("%d ", matrix[i][left]);
            }
            left++;
        }
    }
}

void prefix_sum(int rows, int cols, int matrix[rows][cols], int prefix[rows][cols]) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            prefix[i][j] = matrix[i][j];
            if (i > 0) prefix[i][j] += prefix[i - 1][j];
            if (j > 0) prefix[i][j] += prefix[i][j - 1];
            if (i > 0 && j > 0) prefix[i][j] -= prefix[i - 1][j - 1];
        }
    }
}

int main() {
    int prefix[3][3];

    printf("Original Matrix:\n");
    print_matrix(3, 3, matrix);

    printf("\nPrefix Sum Matrix:\n");
    prefix_sum(3, 3, matrix, prefix);
    print_matrix(3, 3, prefix);

    printf("\nTransposed Matrix:\n");
    transpose(3, 3, matrix);
    print_matrix(3, 3, matrix);

    printf("\nRotated 90 Degrees:\n");
    rotate_90(3, 3, matrix);
    print_matrix(3, 3, matrix);

    printf("\nRotated 180 Degrees:\n");
    rotate_180(3, 3, matrix);
    print_matrix(3, 3, matrix);

    printf("\nSpiral Print:\n");
    spiral_print(3, 3, matrix);
    printf("\n");

    return 0;
}

